system/lib/wasmfs/file.h

```diff
@@ -177,6 +178,9 @@ class Directory : public File {
   struct DCacheEntry {
     DCacheKind kind;
     std::shared_ptr<File> file;
+#ifdef WASMFS_CASE_INSENSITIVE
+    std::string originalName;  // Used for case preservation.
```

Comment:
Case preservation is limited by in-memory cache and can be removed if it looks bad. But for me this looks better than nothing.